ot/mpz-ot/src/chou_orlandi/sender.rs

```diff
@@ -0,0 +1,201 @@
+        }
+    }
+
+    /// Sets up the Sender.
+    ///
+    /// # Arguments
+    ///
+    /// * `sink` - The sink to send messages to the receiver
+    /// * `stream` - The stream to receive messages from the receiver
+    pub async fn setup<Si: IoSink<Message> + Send + Unpin, St: IoStream<Message> + Send + Unpin>(
+        &mut self,
+        sink: &mut Si,
+        stream: &mut St,
+    ) -> Result<(), SenderError> {
+        let sender = self.state.replace(State::Error).into_initialized()?;
+
+        // If the receiver is committed, we run the cointoss protocol
+        if sender.config().receiver_commit() {
+            self.cointoss_receiver = Some(execute_cointoss(sink, stream).await?);
+        }
+
+        let (msg, sender) = sender.setup();
+
+        sink.send(Message::SenderSetup(msg)).await?;
+
+        self.state = State::Setup(sender);
+
+        Ok(())
+    }
+}
+
+async fn execute_cointoss<
+    Si: IoSink<Message> + Send + Unpin,
+    St: IoStream<Message> + Send + Unpin,
+>(
+    sink: &mut Si,
+    stream: &mut St,
+) -> Result<cointoss::Receiver<cointoss::receiver_state::Received>, SenderError> {
+    let receiver = cointoss::Receiver::new(vec![thread_rng().gen()]);
+
+    let commitment = stream
+        .expect_next()
+        .await?
+        .into_cointoss_sender_commitment()?;
+
+    let (receiver, payload) = receiver.reveal(commitment)?;
+
+    sink.send(Message::CointossReceiverPayload(payload)).await?;
+
+    Ok(receiver)
+}
+
+impl ProtocolMessage for Sender {
+    type Msg = Message;
+}
+
+#[async_trait]
+impl OTSender<[Block; 2]> for Sender {
+    async fn send<Si: IoSink<Message> + Send + Unpin, St: IoStream<Message> + Send + Unpin>(
+        &mut self,
+        sink: &mut Si,
+        stream: &mut St,
+        input: &[[Block; 2]],
+    ) -> Result<(), OTError> {
+        let mut sender = self
+            .state
+            .replace(State::Error)
+            .into_setup()
+            .map_err(SenderError::from)?;
+
+        let receiver_payload = stream
+            .expect_next()
+            .await?
+            .into_receiver_payload()
+            .map_err(SenderError::from)?;
+
+        let input = input.to_vec();
+        let (sender, payload) = Backend::spawn(move || {
+            let payload = sender.send(&input, receiver_payload);
+            (sender, payload)
+        })
+        .await;
+
+        let payload = payload.map_err(SenderError::from)?;
+
+        sink.send(Message::SenderPayload(payload)).await?;
+
+        self.state = State::Setup(sender);
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl VerifiableOTSender<bool, [Block; 2]> for Sender {
+    async fn verify_choices<
+        Si: IoSink<Message> + Send + Unpin,
+        St: IoStream<Message> + Send + Unpin,
+    >(
+        &mut self,
+        _sink: &mut Si,
+        stream: &mut St,
+    ) -> Result<Vec<bool>, OTError> {
+        let sender = self
+            .state
+            .replace(State::Complete)
```

Comment:
this should be State::Error. State::Complete should be set at the end of this fn